### PR TITLE
MRG, ENH: Add fine calibration calculation

### PR DESCRIPTION
**Pull request by larsoner**
Coming back to #2818 with a new set of eyes (four years later!) I can get pretty close to replicating MaxFilter's performance, but not quite there. Still need to:

- [x] fix the grad imbalance calculations
- [x] add test files
- [x] add tests

### CODE REVIEW

**Comment by hoechenberger** [comment]
Holy smokes, I didn't even know this was possible! **SICK** 😁

**Comment by larsoner** [comment]
Okay it's working now, ready for review/merge from my end

File: mne/preprocessing/_fine_cal.py [hunk 1]
```diff
@@ -3,9 +3,421 @@
...
+
+    1. Estimate magnetometer normal directions and scale factors. All
+       coils (mag and matching grad) are rotated by the adjusted normal
+       direction.
+    2. Estimate gradiometer imbalance factors. These add point magnetometers
+       in just the gradiometer difference direction or in all three directions
+       (depending on ``n_imbalance``).
+
+    Step (1) is typically the most time consuming step, so step (2) can be
+    iteratively performed using 1 or 3 directions by passing ``calibration``
```
Comment (agramfort) [inline mne/preprocessing/_fine_cal.py:78]:
```suggestion
    iteratively performed using 1 or 3 directions by passing to ``calibration``
```

? I don't get the sentence

File: mne/preprocessing/_fine_cal.py [hunk 2]
```diff
@@ -3,9 +3,421 @@
...
+    """Get the angle between two vectors in degrees."""
+    return np.abs(np.arccos(
+        np.clip((x * y).sum(axis=-1) /
+                (np.linalg.norm(x, axis=-1) *
+                 np.linalg.norm(y, axis=-1)), -1, 1.)))
+
+
+def _adjust_mag_normals(info, data, origin, ext_order):
+    """Adjust coil normals using magnetometers and empty-room data."""
+    from scipy.optimize import fmin_cobyla
```
Comment (agramfort) [inline mne/preprocessing/_fine_cal.py:217]: neat !

File: mne/preprocessing/_fine_cal.py [hunk 3]
```diff
@@ -21,15 +433,14 @@ def read_fine_calibration(fname):
 
     Returns
     -------
-    calibration : dict
-        Fine calibration information.
+    fine_cal : dict
+        Fine calibration information. Each channel (key) in ``calibration`` is
+        a dict with ``'loc'`` and ``'imb'`` keys.
```
Comment (agramfort) [inline mne/preprocessing/_fine_cal.py:438]: can you recall here what loc and imb mean what is there shapes?

File: mne/preprocessing/_fine_cal.py [hunk 4]
```diff
@@ -49,36 +460,28 @@ def read_fine_calibration(fname):
...
-    locs = np.array(locs)
-    return dict(ch_names=ch_names, locs=locs, imb_cals=imb_cals)
+            cal[ch_name] = dict(
+                loc=np.array([float(x) for x in vals[1:13]]),
+                imb=np.array([float(x) for x in vals[13:]]))
+    return cal
 
 
-def write_fine_calibration(fname, calibration):
+def write_fine_calibration(fname, fine_cal):
```
Comment (agramfort) [inline mne/preprocessing/_fine_cal.py:471]: API change but I doubt it will be a pb... I am getting old...

**Comment by larsoner** [comment]
Pushed fixes that hopefully address your comments @agramfort

File: mne/preprocessing/_fine_cal.py [hunk 5]
```diff
@@ -3,9 +3,424 @@
...
+       direction.
+    2. Estimate gradiometer imbalance factors. These add point magnetometers
+       in just the gradiometer difference direction or in all three directions
+       (depending on ``n_imbalance``).
+
+    Magnetometer normal and coefficient estimation (1) is typically the most
+    time consuming step. Gradiometer imbalance parameters (2) can be
+    iteratively reestimated (for example, first using ``n_imbalance=1`` then
+    subsequently ``n_imbalance=3``) by passing the previous ``calibration``
+    output to the ``calibration`` input in the second call.
```
Comment (larsoner) [inline mne/preprocessing/_fine_cal.py:81]: @agramfort I rewrote these to clarify the purpose and use case

**Review by agramfort — APPROVED** [review]

**Pull request merged**